### PR TITLE
[Serve] Document how to stream outputs from `@serve.batch`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

#35729 and  #36136 let users stream outputs from `@serve.batch`-decorated functions. This change adds documentations explaining how to use this feature.

Updated doc link: https://anyscale-ray--36263.com.readthedocs.build/en/36263/serve/performance.html#streaming-batched-requests

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [X] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
Looks good, but we should also add a more practical example with text generation. Are you planning to do that separately?

**Comment by shrekris-anyscale** [comment]
Yeah, I think it would be better to do it separately. I'm planning to add a doc to the `Examples` section.

File: doc/source/serve/doc_code/batching_guide.py [hunk 1]
```diff
@@ -33,3 +34,65 @@ async def __call__(self, multiple_samples: List[int]) -> List[int]:
...
+
+from ray import serve
+
+
+@serve.deployment
+class StreamingResponder:
+    async def generate_numbers(self, max: int) -> AsyncGenerator[int, None]:
+        for i in range(max):
+            yield i
+            asyncio.sleep(0.1)
```
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:53]: missing `await`

File: doc/source/serve/doc_code/batching_guide.py [hunk 2]
```diff
@@ -33,3 +34,65 @@ async def __call__(self, multiple_samples: List[int]) -> List[int]:
...
+
+    def __call__(self, request: Request) -> StreamingResponse:
+        max = int(request.query_params.get("max", "25"))
+        gen = self.generate_numbers(max)
+        return StreamingResponse(gen, status_code=200, media_type="text/plain")
+
+
+# __single_stream_end__
+
+# TODO (shrekris-anyscale): add unit tests
```
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:63]: please add tests here before merging

File: doc/source/serve/doc_code/batching_guide.py [hunk 3]
```diff
@@ -33,3 +34,65 @@ async def __call__(self, multiple_samples: List[int]) -> List[int]:
...
+    ) -> AsyncGenerator[List[int], None]:
+        for i in range(max(max_list)):
+            next_numbers = []
+            for requested_max in max_list:
+                if requested_max > i:
+                    next_numbers.append(i)
+                else:
+                    next_numbers.append(StopIteration)
+            yield next_numbers
+            asyncio.sleep(0.1)
```
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:88]: missing await

File: doc/source/serve/doc_code/batching_guide.py [hunk 4]
```diff
@@ -33,3 +34,65 @@ async def __call__(self, multiple_samples: List[int]) -> List[int]:
...
+        self, max_list: List[int]
+    ) -> AsyncGenerator[List[int], None]:
+        for i in range(max(max_list)):
+            next_numbers = []
+            for requested_max in max_list:
+                if requested_max > i:
+                    next_numbers.append(i)
+                else:
+                    next_numbers.append(StopIteration)
+            yield next_numbers
```
Comment (edoakes) [inline doc/source/serve/doc_code/batching_guide.py:87]:
```suggestion
            yield [
                i for i, requested_max in enumerate(max_list) if i <= requested_max else StopIteration
            ]
```
might be easier to read? not sure

File: doc/source/serve/performance.md [hunk 5]
```diff
@@ -121,6 +121,30 @@ You can supply two optional parameters to the decorators.
 - `max_batch_size` controls the size of the batch.
 Once the first request arrives, the batching decorator will wait for a full batch (up to `max_batch_size`) until `batch_wait_timeout_s` is reached. If the timeout is reached, the batch will be sent to the model regardless the batch size.
 
+### Streaming batched requests
+
+You can stream the outputs from your batched requests with an async generator. Let's convert this `StreamingResponder` class to accept a batch.
```
Comment (edoakes) [inline doc/source/serve/performance.md:126]:
```suggestion
Use an async generator to stream the outputs from your batched requests. Let's convert the `StreamingResponder` class to accept a batch.
```

File: doc/source/serve/performance.md [hunk 6]
```diff
@@ -121,6 +121,30 @@ You can supply two optional parameters to the decorators.
...
+
+```{literalinclude} doc_code/batching_guide.py
+---
+start-after: __single_stream_begin__
+end-before: __single_stream_end__
+---
+```
+
+You can decorate async generator functions with the {mod}`ray.serve.batch` decorator. Similar to non-streaming methods, the function should take in a `List` of inputs. In each iteration, it should `yield` an iterable of outputs with the same length as the input batch size.
```
Comment (edoakes) [inline doc/source/serve/performance.md:135]:
```suggestion
Decorate async generator functions with the {mod}`ray.serve.batch` decorator. Similar to non-streaming methods, the function takes in a `List` of inputs and in each iteration it `yield`s an iterable of outputs with the same length as the input batch size.
```

**Review by edoakes — APPROVED** [review]
LGTM pending nits

File: doc/source/serve/doc_code/batching_guide.py (hunk 1 above)
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: Thanks, I added the `await`.

File: doc/source/serve/doc_code/batching_guide.py (hunk 2 above)
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: I added some tests.

File: doc/source/serve/doc_code/batching_guide.py (hunk 3 above)
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: Thanks, I added the `await`.

File: doc/source/serve/doc_code/batching_guide.py (hunk 4 above)
Comment (shrekris-anyscale) [reply doc/source/serve/doc_code/batching_guide.py]: This feels harder to read in my opinion. I'd prefer to stick with the explicit `if`-`else`.

**Review by angelinalg — APPROVED** [review]

**Pull request merged**